src/internal/i18n/provider.tsx

```diff
@@ -0,0 +1,147 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import React, { useCallback, useContext, useMemo } from 'react';
+import IntlMessageFormat from 'intl-messageformat';
+import { MessageFormatElement } from '@formatjs/icu-messageformat-parser';
+
+import { InternalI18nContext, FormatFunction, CustomHandler } from './context';
+
+export interface I18nProviderProps {
+  messages: ReadonlyArray<I18nProviderProps.Messages>;
+  locale?: string;
+  children: React.ReactNode;
+}
+
+export namespace I18nProviderProps {
+  export interface Messages {
+    [namespace: string]: {
+      [locale: string]: {
+        [component: string]: {
+          [key: string]: string | MessageFormatElement[];
+        };
+      };
+    };
+  }
+}
+
+/**
+ * Context to send parent messages down to child I18nProviders. This isn't
+ * included in the InternalI18nContext to avoid components from depending on
+ * MessageFormatElement types.
+ */
+const I18nMessagesContext = React.createContext<I18nProviderProps.Messages>({});
+
+export function I18nProvider({ messages: messagesArray, locale: providedLocale, children }: I18nProviderProps) {
+  // Keep a cache per render for improved performance. Generally, it's expected
+  // to have an I18nProvider pretty high up the render tree and for it to
+  // rarely rerender, so it should stick around and offer performance gains.
+  const cache = useMemo(() => new Map<string, unknown>(), []);
+  cache.clear();
+
+  // The provider accepts an array of configs. We merge parent messages and
+  // flatten the tree early on so that accesses by key are simpler and faster.
+  const parentMessages = useContext(I18nMessagesContext);
+  const messages = useMemo(() => mergeMessages([parentMessages, ...messagesArray]), [parentMessages, messagesArray]);
```

Comment:
This memo is useless, as in all usage examples before, you pass a new array on every call

```
<I18nProvider messages={[MESSAGES]} locale="en"></I18nProvider>
```